gapis/api/vulkan/query_timestamps.go

```diff
@@ -0,0 +1,437 @@
+		cmd.Extras().Observations().ApplyWrites(s.Memory.ApplicationPool())
+		out.MutateAndWrite(ctx, id, cmd)
+		info := cmd.pProperties.MustRead(ctx, cmd, s, nil)
+		t.timestampPeriod = info.Limits().TimestampPeriod()
+		log.I(ctx, "TimestampPeriod is %v", t.timestampPeriod)
+
+	case *VkCreateDevice:
+		out.MutateAndWrite(ctx, id, cmd)
+		cmd.Extras().Observations().ApplyWrites(s.Memory.ApplicationPool())
+		t.device = cmd.pDevice.MustRead(ctx, cmd, s, nil)
+
+	case *VkQueueSubmit:
+		cmd.Extras().Observations().ApplyReads(s.Memory.ApplicationPool())
+		queue := cmd.Queue()
+		queueFamilyIndex := GetState(s).Queues().Get(queue).Family()
+		submitCount := cmd.SubmitCount()
+		submitInfos := cmd.pSubmits.Slice(0, uint64(submitCount), s.MemoryLayout).MustRead(ctx, cmd, s, nil)
+		cmdBufferCount := uint32(0)
+		for i := uint32(0); i < submitCount; i++ {
+			si := submitInfos[i]
+			cmdBufferCount += si.CommandBufferCount()
+		}
+		queryCount := cmdBufferCount * 2
+
+		t.createComandpoolIfNeeded(ctx, cb, out, queueFamilyIndex)
+		t.createQueryPoolIfNeeded(ctx, cb, out, queryCount)
+		t.rewriteQueueSubmit(ctx, cb, out, id, cmd)
+
+		waitCmd := cb.VkQueueWaitIdle(queue, VkResult_VK_SUCCESS)
+		out.MutateAndWrite(ctx, id, waitCmd)
+
+		// Get result back.
+		buflen := uint64(queryCount * 8)
+		tmp := s.AllocOrPanic(ctx, buflen)
+		flags := VkQueryResultFlags(VkQueryResultFlagBits_VK_QUERY_RESULT_64_BIT | VkQueryResultFlagBits_VK_QUERY_RESULT_WAIT_BIT)
+		newCmd := cb.VkGetQueryPoolResults(
+			t.device,
+			t.queryPool,
+			0,
+			queryCount,
+			memory.Size(buflen),
+			tmp.Ptr(),
+			8,
+			flags,
+			VkResult_VK_SUCCESS)
+
+		out.MutateAndWrite(ctx, api.CmdNoID, newCmd)
+
+		out.MutateAndWrite(ctx, api.CmdNoID, cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
+			b.ReserveMemory(tmp.Range())
+			b.Post(value.ObservedPointer(tmp.Address()), buflen, func(r binary.Reader, err error) {
+				if err != nil {
+					log.I(ctx, "b post get err %v", err)
+					return
+				}
+
+				tStart := r.Uint64()
+				for i := uint32(1); i < queryCount; i++ {
+					tEnd := r.Uint64()
+					t.timestamps[t.resultIndex].Time = uint64(float32(tEnd-tStart) * t.timestampPeriod)
+					t.resultIndex++
+					tStart = tEnd
+				}
+
+			})
+			return nil
+		}))
+
+		tmp.Free()
+
+	case *VkDestroyCommandPool:
+		cmd.Extras().Observations().ApplyReads(s.Memory.ApplicationPool())
+		writeEach(ctx, out,
+			cb.VkDestroyCommandPool(
+				t.device,
+				t.commandPool,
+				memory.Nullptr),
+			cb.VkDestroyQueryPool(
+				t.device,
+				t.queryPool,
+				memory.Nullptr))
+
+		out.MutateAndWrite(ctx, id, cmd)
+	default:
+		out.MutateAndWrite(ctx, id, cmd)
+	}
+}
+
+func (t *queryTimestamps) Flush(ctx context.Context, out transform.Writer) {
+	s := out.State()
+	cb := CommandBuilder{Thread: 0, Arena: s.Arena}
+	out.MutateAndWrite(ctx, api.CmdNoID, cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
```

Comment:
We shouldn't need this for this transform. This was to make sure that a report actually finished (so we could catch crashes / validation layer warnings) but in this case, we can just wait for all of the posts we are expecting.

Reply:
As discussed, this will be changed later.